### DISCUSSION TITLE
Roundtripping png to pdf

### DISCUSSION THREAD

**Discussion by Jmuccigr** [ImageMagick/ImageMagick#3010] [Help]
I realize that working with images in PDF can be a bit tricky sometimes, but I was hoping for an answer about something I've observed.

If I take a png and convert it with IM to a pdf, I sometimes end up with a PDF that's smaller than the png. Using pdfimages to get the png back out will give me more or less the originally sized image and the extracted image also appears to be identical to the original, even if the file size differs a bit.

I attach an example png. If I convert it to pdf via:

`magick orig.png demo.pdf`

the 458k png becomes a 379k PDF. Extracting the png gives me a 452k file:

`pdfimages -all demo.pdf demo`

I know that sticking an image in a PDF removes some of the metadata like resolution, but does that account for **all** the lost size? `identify` shows that a few chunks are missing from the pdfimages png (demo) compared with the original. I'm also curious how the PDF loses that much in size.

![orig](https://user-images.githubusercontent.com/615115/102702379-008c7300-4230-11eb-9037-38afd8eafd85.png)

**Comment by snibgo** [comment]
IM optimizes PNG only in simple ways, eg grayscale and palette. The user can select other options at affect compression, but IM won't try them to find the best.

Other programs such as pngcrush are better for that. Your PNG is 458405 bytes. Using:
```
pngcrush forpdf.png x.png
```
... with no fancy options makes x.png with 371230 bytes, a significant improvement. IM says the pixels in forpdf.png and x.png are identical, of course.

When converting these PNG files to PDFs, the PDFs are virtually the same size. This suggests that IM is optimizing when writing to PDF more than when it writes to PNG. Perhaps IM could use the same method to optimize PNG files, or at least provide that as an option.

**Reply by Jmuccigr** [reply]
> When converting these PNG files to PDFs, the PDFs are virtually the same size. This suggests that IM is optimizing when writing to PDF more than when it writes to PNG. Perhaps IM could use the same method to optimize PNG files, or at least provide that as an option.

Yes, this is exactly what I mean. IM is doing something when it creates the pdf and I'd love to know what that is. It makes a smaller file (at least in this case) and then, whatever it is, pdfimages undoes it when it extracts the image, creating a file with the same pixels as the original.

**Reply by Jmuccigr** [reply]
PS If I change the workflow to use pngcrush, here's what I get:

`pngcrush orig.png crush.png` results in a 344kb file (not the 371kb @snibgo got above)
`magick crush.png crush.pdf` results in a 379kb file, almost exactly the same size as the PDF originally made by IM.

So maybe IM is running pngcrush or something equivalent before making the PDF?

**Reply by Jmuccigr** [reply]
For fun I converted the original jpg to png using all possible values of compression filter, level, and strategy, but I can't get down to the pngcrush size. The best I can do is by setting those values, respectively, to 130, 131, 230, 231, 330, 331, 430, or 431, all of which give identically sized files: 357630. The pngcrush output was 343905.

Attaching the original jpg.
![orig](https://user-images.githubusercontent.com/615115/102817288-824ddf00-439d-11eb-9730-3705de89b845.jpg)

**Reply by fmw42** [reply]
As user snibgo pointed out, IM compressed png in very simple ways and will not achieve the better compression of the specialized pngcrush.

**Reply by Jmuccigr** [reply]
> As user snibgo pointed out, IM compressed png in very simple ways and will not achieve the better compression of the specialized pngcrush.

OK, but the point is that it made it much smaller when it put the png in the pdf, so how did it manage that?

**Reply by fmw42** [reply]
Perhaps when putting any file into PDF, it is converted to PAM or PNM format or some other compressed format. Or ImageMagick may call pngcrush to put the png into the PDF? I do not know. Only viewing the code or getting a response from the IM developers will tell you.

**Reply by Jmuccigr** [reply]
> getting a response from the IM developers will tell you.

Well, I was hoping that this kind of post would lead to that. 😄
